mockgcp/mockcontaineranalysis/note.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mockcontaineranalysis
+
+import (
+	"context"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/common/projects"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	pb "github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/generated/mockgrafeas/v1"
+	"github.com/golang/protobuf/ptypes/empty"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+type NoteName struct {
+	Project  *projects.ProjectData
+	NoteName string
+}
+
+type ContainerAnalysisV1 struct {
+	*MockService
+	pb.UnimplementedGrafeasServer
+}
+
+func (s *ContainerAnalysisV1) GetNote(ctx context.Context, req *pb.GetNoteRequest) (*pb.Note, error) {
+	name, err := s.parseNoteName(req.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+
+	obj := &pb.Note{}
+	if err := s.storage.Get(ctx, fqn, obj); err != nil {
+		return nil, err
+	}
+
+	return obj, nil
+}
+
+func (s *ContainerAnalysisV1) CreateNote(ctx context.Context, req *pb.CreateNoteRequest) (*pb.Note, error) {
+	reqName := req.Parent + "/notes/" + req.NoteId
+	name, err := s.parseNoteName(reqName)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
```

Comment:
I think we want to assign the note fields according to https://cloud.google.com/artifact-analysis/docs/reference/rest/v1beta1/projects.notes#Note Specifically I don't see field like `sbomReference`  `vulnerabilityAssessment`, etc

Reply:
I'm unsure if they are required. 

Checking the [http_log](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/9f66475fcaa28abd8bfceec1035e2533668601bf/pkg/test/resourcefixture/testdata/basic/containeranalysis/v1beta1/containeranalysisnote/_http.log), fields are assigned according to the [test yaml](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/test/resourcefixture/testdata/basic/containeranalysis/v1beta1/containeranalysisnote/create.yaml). `sbomReference` and `vulnerabilityAssessment` are both options of union type, per the REST API doc, `// Union field type can be only one of the following`, once `attestation` is set, we may not need other types. 

However, I checked other mock resources, missing some field values might cause mock test to fail. So I'm unsure if we have an "official" way to determine the necessarily of the manual assigned values.

Reply:
I didn't make it clear. 
1. I think we want to assign the output only fields here.  like "kind"
2. Regarding `vulnerabilityAssessment` and `sbomRefernece`, I saw they are one of "Union field type can be only one of the following". I want to know what error CREATE will give if specify two, and what message the UPDATE will give if we change the one of from sbomRefrence to valnerabilityAssessment, etc.  These can be done in follow up PRs.

Reply:
Make sense! Done.